client/allocrunner/alloc_runner.go

```diff
@@ -729,14 +729,35 @@ func (ar *allocRunner) killTasks() map[string]*structs.TaskState {
 	// run alloc prekill hooks
 	ar.preKillHooks()
 
+	// generate task event for given task runner
+	taskEventFn := func(tr *taskrunner.TaskRunner) (te *structs.TaskEvent) {
+		te = structs.NewTaskEvent(structs.TaskKilling).
+			SetKillTimeout(tr.Task().KillTimeout, ar.clientConfig.MaxKillTimeout)
+
+		// if the task is not set failed, the job type is batch,
+		// and the allocation is being migrated then update the
+		// kill reason and mark the task as failed. this ensures
+		// the task is recreated if no eligible nodes are immediately
+		// available.
+		if !tr.TaskState().Failed &&
+			ar.alloc.Job.Type == structs.JobTypeBatch &&
+			ar.alloc.DesiredTransition.Migrate != nil &&
+			*ar.alloc.DesiredTransition.Migrate {
```

Comment:
In the case where we haven't reach the drain deadline, this won't be set yet, right? So batch allocs that do manage to finish on time won't hit this case.

Reply:
Thank you for calling this out. Originally, no, the state would still get failed if the task had already completed since the `Kill` is called regardless. Small update includes a check for if the task is completed before marking and updated test includes verification that only still running tasks are marked as failed.